nav2_costmap_2d/plugins/plugin_container_layer.cpp

```diff
@@ -0,0 +1,218 @@
+
+using std::vector;
+
+namespace nav2_costmap_2d
+{
+PluginContainerLayer::PluginContainerLayer()
+: default_plugins_{},
+  default_types_{
+    "nav2_costmap_2d::StaticLayer",
+    "nav2_costmap_2d::ObstacleLayer",
+    "nav2_costmap_2d::InflationLayer"}
+{
+  costmap_ = NULL;
+}
+
+PluginContainerLayer::~PluginContainerLayer()
+{
+  costmap_ = NULL;
+}
+
+void PluginContainerLayer::onInitialize()
+{
+  auto node = node_.lock();
+
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  combined_costmap_.setDefaultValue(0);
+
+  node->declare_parameter(name_ + "." + "enabled", rclcpp::ParameterValue(true));
+  node->declare_parameter(name_ + "." + "plugins", rclcpp::ParameterValue(default_plugins_));
+  node->declare_parameter(name_ + "." + "combination_method", rclcpp::ParameterValue(1));
+
+  node->get_parameter(name_ + "." + "enabled", enabled_);
+  node->get_parameter(name_ + "." + "plugins", plugin_names_);
+  node->get_parameter(name_ + "." + "combination_method", combination_method_);
+
+  plugin_types_.resize(plugin_names_.size());
+
+  for (size_t i = 0; i < plugin_names_.size(); ++i) {
+    plugin_types_[i] = nav2_util::get_plugin_type_param(node, name_ + "." + plugin_names_[i]);
+  }
+
+  // Then load and add the plug-ins to the costmap
+  for (unsigned int i = 0; i < plugin_names_.size(); ++i) {
+    std::shared_ptr<Layer> plugin = plugin_loader_.createSharedInstance(plugin_types_[i]);
+    std::unique_lock<Costmap2D::mutex_t> lock(*(layered_costmap_->getCostmap()->getMutex()));
+    addPlugin(plugin, name_ + "." + plugin_names_[i]);
+    lock.unlock();
+  }
+
+  PluginContainerLayer::matchSize();
+  current_ = true;
+}
+
+void PluginContainerLayer::addPlugin(std::shared_ptr<Layer> plugin, std::string layer_name)
+{
+  plugins_.push_back(plugin);
+  auto node = node_.lock();
+  plugin->initialize(layered_costmap_, name_ + "." + layer_name, tf_, node, callback_group_);
+}
+
+void PluginContainerLayer::updateBounds(
+  double robot_x,
+  double robot_y,
+  double robot_yaw,
+  double * min_x,
+  double * min_y,
+  double * max_x,
+  double * max_y)
+{
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->updateBounds(robot_x, robot_y, robot_yaw, min_x, min_y, max_x, max_y);
+  }
+}
+
+void PluginContainerLayer::updateCosts(
+  nav2_costmap_2d::Costmap2D & master_grid,
+  int min_i,
+  int min_j,
+  int max_i,
+  int max_j)
+{
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
+  if (!enabled_) {
+    return;
+  }
+
+  combined_costmap_.resetMap(min_i, min_j, max_i, max_j);
+
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->updateCosts(combined_costmap_, min_i, min_j, max_i, max_j);
+  }
+
+  costmap_ = combined_costmap_.getCharMap();
+
+  switch (combination_method_) {
```

Comment:
I think you're missing some https://github.com/ros-navigation/navigation2/blob/a0c11b441b5faa1924ba31953900e14cc0a055f1/nav2_costmap_2d/plugins/obstacle_layer.cpp#L552-L564